public/app/plugins/datasource/azuremonitor/types/types.ts

```diff
@@ -14,86 +14,45 @@ import Datasource from '../datasource';
 import { AzureLogAnalyticsMetadataTable } from './logAnalyticsMetadata';
 import { AzureMonitorQuery, ResultFormat } from './query';
 
-export type AzureDataSourceSettings = DataSourceSettings<AzureDataSourceJsonData, AzureDataSourceSecureJsonData>;
-export type AzureDataSourceInstanceSettings = DataSourceInstanceSettings<AzureDataSourceJsonData>;
-
-export interface DatasourceValidationResult {
-  status: 'success' | 'error';
-  message: string;
-  title?: string;
-}
-
-/**
- * Azure clouds known to Azure Monitor.
- */
-export enum AzureCloud {
-  Public = 'AzureCloud',
-  China = 'AzureChinaCloud',
-  USGovernment = 'AzureUSGovernment',
-  None = '',
-}
-
-export type AzureAuthType = 'msi' | 'clientsecret' | 'workloadidentity';
-
-export type ConcealedSecret = symbol;
-
-interface AzureCredentialsBase {
-  authType: AzureAuthType;
-}
+type DataSourceJsonValue =
+  | string
+  | number
+  | boolean
+  | null
+  | undefined
+  | { [x: string]: DataSourceJsonValue }
+  | DataSourceJsonValue[];
 
-export interface AzureManagedIdentityCredentials extends AzureCredentialsBase {
-  authType: 'msi';
+export interface DataSourceJsonData2 extends DataSourceJsonData {
```

Comment:
This is not a solution that I'd be happy to merge. Please see my original comment [here](https://github.com/grafana/grafana/pull/79932#issuecomment-1927867463). I would rather stick to the base types and extend a custom type that is more specific over a type where any property can be specified.

Reply:
I would not want to have credentials structure directly accessible from the datasource's business layer:
- Fields of `azureCredentials` should not be accessible directly but only through `getCredentials()` and other functions of SDK. If such fields be accessible directly it will be temptation to do shortcuts and read them directly, e.g. cloud name etc.
- Different variations of legacy fields which would just pollute schema.

JsonData is a property bag, it should be typed as such in Grafana Plugin SDK.

Reply:
I think to not block this I'd rather expose the credentials in the data source layer. You can create a PR to `@grafana/data` to update the types but this solution isn't a nice one that I'd like to merge.

I'd rather include the various variations of the legacy fields and temporarily pollute the schema until the base problem is solved. It will make the code clearer in the short term whilst also not attempting to overwrite the base types.

Reply:
Changed to strongly-typed interface.